### DISCUSSION TITLE
Compatibility with WSL(2)?

### DISCUSSION THREAD

**Discussion by jpfeuffer** [microsoft/vscode-cmake-tools#2688] [Q&A] [answered]
Hi,

I have seen a lot of issues that were closed and people were referred to the "Remote: WSL" extension.
I tried that extension but I think the cmake-tools plugin is just not compatible at all.
Here are my problems:

1) It won't even install on the remote instance that spawns with `code .` unless you declare it as a UI plugin in your settings.
2) It does not find the compilers/kits on the WSL instance (I installed gcc/++ with apt), no matter what I do. Neither in the Windows instance nor on the Remote WSL instance.
3) Extended configuration is almost impossible, since Workspace configurations overwrite the Remote WSL configurations
4) Even if I move everything to User configuration and overwrite with the Remote WSL configuration, everything is full of `[proc] The command: /usr/bin/cmake --version failed with error: Error: spawn /usr/bin/cmake ENOENT` errors, while of course running `/usr/bin/cmake --version` from the terminal of the same Remote WSL instance works flawlessly.

TLDR: How to use this extension with WSL? All I want to do is to build the same project in the same folder once with MSVC and once with GCC WSL, preferably even without having two VS Code windows open. Is there any tutorial or documentation?

**Comment by elahehrashedi** [comment]
Question 1: There shouldn't be an issue installing the extension on WSL, regardless of how you open a WSL window:

![image](https://user-images.githubusercontent.com/16143955/184209620-c05b1f72-bbc1-4bf9-9305-75eea5f8f091.png)

Would you explain more about what kind of error you encounter when trying to install the extension?

Question 2: Can you run `"gcc -v"` and share the results with us?
I am not sure why the extension can not pick up the kit. Can you try and add the kit manually to the `.vscode/cmake-kits.json` file and see if you can get that to work, here is an example:
```json
  {
    "name": "GCC 7.5.0 x86_64-linux-gnu",
    "compilers": {
      "C": "/usr/bin/gcc-7",
      "CXX": "/usr/bin/g++-7"
    }
```

**Reply by jpfeuffer** [reply]
Hi!

But you have the same error. See your yellow exclamation mark.

**Reply by jpfeuffer** [reply]
```
user@DESKTOP:/mnt/c/git/project$ g++ --version
g++ (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Reply by jpfeuffer** [reply]
```
> g++ -v
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/9/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:hsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 9.4.0-1ubuntu1~20.04.1' --with-bugurl=file:///usr/share/doc/gcc-9/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,gm2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-9 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-9-Av3uEd/gcc-9-9.4.0/debian/tmp-nvptx/usr,hsa --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 9.4.0 (Ubuntu 9.4.0-1ubuntu1~20.04.1) 
```

**Reply by elahehrashedi** [reply]
![image](https://user-images.githubusercontent.com/16143955/184211958-ea61069e-75b1-4845-adb3-6b4a2717b5bd.png)
That is not an error. Are you able to install the extension by clicking on "Install in WSL: ..." ? This will install the extension on your emote machine (this is a UI extension, read more : https://code.visualstudio.com/api/advanced-topics/remote-extensions#architecture-and-extension-types)

**Reply by elahehrashedi** [reply]
seems like you have a working compiler, can you try manually adding the kit?

**Reply by jpfeuffer** [reply]
When I use:
```
  {
    "name": "GCC9 WSL",
    "preferredGenerator": {
      "name": "Ninja"
    },
    "compilers": {
      "C": "/usr/bin/gcc",
      "CXX": "/usr/bin/g++"
    },
    "keep": true
  },
```
and select this Kit in the Windows instance:

```
[proc] Executing command: "C:\Program Files\CMake\bin\cmake.EXE" --no-warn-unused-cli -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -DCMAKE_BUILD_TYPE:STRING=Debug -DCMAKE_C_COMPILER:FILEPATH=/usr/bin/gcc -DCMAKE_CXX_COMPILER:FILEPATH=/usr/bin/g++ -Sc:/git/project "-Bc:/git/project/cmake-build-Debug-GCC9 WSL" -G Ninja
[cmake] Not searching for unused variables given on the command line.
[cmake] -- The C compiler identification is unknown
[cmake] -- The CXX compiler identification is unknown
[cmake] CMake Error at CMakeLists.txt:43 (project):
[cmake]   The CMAKE_C_COMPILER:
[cmake] 
[cmake]     /usr/bin/gcc
[cmake] 
[cmake]   is not a full path to an existing compiler tool.

```

**Reply by jpfeuffer** [reply]
For the WSL instance, I get right away:
```
CMakeLists.txt was not found in the root of the folder "project". How would you like to proceed?
```

**Reply by jpfeuffer** [reply]
sourceDirectory is of course `${workspaceFolder}` for both instances

**Reply by elahehrashedi** [reply]
> CMakeLists.txt was not found in the root of the folder "project". How would you like to proceed?

You get this message when your workspace folder doesn't have a CMakeLists.txt file. You can proceed by locating or creating one. Make sure you open your project folder, if you already have one. If you are planning to start from scratch, then proceed by creating a new CMakeLists.txt file.

What you are experiencing is by design, not an error. You can use this tutorial to get started with CMake Tools on Linux: https://code.visualstudio.com/docs/cpp/cmake-linux

**Reply by jpfeuffer** [reply]
I am in my project folder. I ran `code .` from my project folder. I can see all my files *including my CMakeLists.txt* in my workspace.
Configuring in the Windows instance also works with Windows compilers (like VS MSVC)

**Reply by jpfeuffer** [reply]
<img width="1008" alt="image" src="https://user-images.githubusercontent.com/8102638/184217615-b0444481-7c71-4c3b-b5cd-325e8e43060d.png">

**Reply by jpfeuffer** [reply]
And my Windows instance of the same workspace:
<img width="1318" alt="image" src="https://user-images.githubusercontent.com/8102638/184218155-fccec9f9-8f32-43fd-a732-b20ab60e338e.png">

**Reply by jpfeuffer** [reply]
I don't even know, why it tries to use the Windows cmake.exe under WSL and everything. It just feels like this extension was not really designed with WSL2 in mind. Are there people actively working with a project where they are using cmake-tools to compile the same project under Windows and WSL2 parallelly?

**Reply by bobbrow** [reply, marked as answer]
Extensions don't generally need to be designed for WSL/WSL2 because the underlying system should be reporting to the extension that it is operating in a linux environment.  It sounds like something is not configured correctly.  I am running the extension in WSL2 and it picks `/usr/bin/cmake` for me.  I tried both a workspace in the linux filesystem and something on `/mnt/d`.

Can you share what the value of `Cmake: CMake Path` is?  I'm wondering if you overrode the setting at some point.
